Support multiple versions of the same API: FulfillmentInboundV0

bdelanghe opened the pull request:
## Summary

  Support multiple versions of the same API (e.g., Fulfillment Inbound V0 alongside 2024-03-20) without clobbering on regeneration. I did a codegen run with these changes on a separate branch to keep things clean, you can see the results [here](https://github.com/bdelanghe/amz_sp_api/pull/5)

## Why

  When Amazon publishes spec updates under the same API module, later codegen runs overwrite earlier ones. This breaks supported legacy only endpoints like V0.
  
## What Changed

 - Conditionally adjust gem and model names for Fulfillment Inbound v0 to allow for the partially deprecated endpoint.

## Testing

  Validated across fulfillment-inbound versions and spot-checked other APIs remain unaffected.

bdelanghe committed:
added beta endpoint for deliveryOfferings

bdelanghe committed:
Refactor Codegen Pipeline: Pull Models, Codegen, Hoist Files and Release

Just run ./run.sh to cut a new release.

bdelanghe committed:
Allow multiple API versions via versioned_models (#4)

ericcj reviewed (commented):
is there a simpler alternative to consider like just keeping the gem on V0?  i don't fully understand why we need both api versions in a single version of the gem?  this is a lot a lot of code and every line of code is another opportunity for bugs.

bdelanghe committed:
Multi version selling partner api models@aad66aa (#6)

bdelanghe committed:
clean (#7)

bdelanghe commented:
> is there a simpler alternative to consider like just keeping the gem on V0? i don't fully understand why we need both api versions in a single version of the gem? this is a lot a lot of code and every line of code is another opportunity for bugs.

@ericcj Just pushed a small, bare-bones change. Amazon has partially deprecated Fulfillment Inbound v0, but some required inbound methods still only exist in v0 and must be used alongside the newer API version.
https://developer-docs.amazon.com/sp-api/docs/fulfillment-inbound-api#considerations

bdelanghe committed:
Codegen Run Against : selling-partner-api-models@aad66aa (#5)

\

bdelanghe commented:
@ericcj I added all of the codegen changes to my main branch. 

I have a draft for the changes since the last version based on PR #30. There has been a lot. I only went as far as version bumps, new domains, and deprecations. I could go over minor updates for models as well, but I figured the release notes are better for specifics

# Updates Since [v1.0.0](https://github.com/ericcj/amz_sp_api/releases/tag/v1.0.0)

This update is based on running codegen on against [amzn/selling-partner-api-models/@aad66aa](https://github.com/amzn/selling-partner-api-models/tree/54f22853bf3937148461ec84a74eb8f58da13056) which was committed on *Tue, 13 Jan 2026*

This was compared against the last cut release after PR #30 [amzn/selling-partner-api-models/@54f2285](https://github.com/amzn/selling-partner-api-models/tree/54f22853bf3937148461ec84a74eb8f58da13056) on which was committed on *Aug 19, 2022*

Needless to say a lot has changed. If you want to see all of the changes look at the official [sp-api-release-notes](https://developer-docs.amazon.com/sp-api/docs/sp-api-release-notes)

### Breaking Changes

- `Authorization v1` & `FBA Small and Light v1` all calls to these domains will need to be removed.
- `productPricing_2022-05-01` replaces the all endpoints with new - no migration guide. Refer to use-case [batch requests](https://developer-docs.amazon.com/sp-api/docs/retrieve-featured-offers-batch-asins)
- `shippingV2` uses many of the same methods refer to the [migration guide](https://developer-docs.shipping.amazon.com/apis/docs/amazon-shipping-migration-guide)
- `fulfillment-inbound-api-model` - all old calls will need to move to `fulfillment-inbound-api-model-v0` for the methods that are still supported. `fulfillment-inbound-api-model` points to all of the new versions endpoints 
- `Data Kiosk API` & `Feeds API` many return datasets and reports are no longer available
- Other Deprecated Methods can be found bellow

## Removed

Nothing is removed. Codegen only creates and modifies existing based on current models. Deprecated and endpoints namespaces are still kept; normally deprecated is replaced by a new version. Refer to [SP-API Deprecations](https://developer-docs.amazon.com/sp-api/docs/sp-api-deprecations) table for more information. The follow endpoints are still available in the gem, but may be currently unsupported or will be soon. 

### Deprecated Domains
- [authorization-api-model/authorization](https://developer-docs.amazon.com/sp-api/docs/sp-api-deprecations#authorization-api)
- [fba-small-and-light-api-model/fbaSmallandLight](https://developer-docs.amazon.com/sp-api/docs/sp-api-deprecations#fba-small-and-light)

### Deprecated Methods & Return Results

- Data Kiosk API: `analytics_salesAndTraffic_2023_11_15` dataset
- Feeds API: [several](https://developer-docs.amazon.com/sp-api/docs/sp-api-deprecations#feeds-api) datasets
- Fulfillment Inbound v0: ([Refer to Guide](https://developer-docs.amazon.com/sp-api/docs/migrating-fulfillment-inbound-workflows))
- Merchant Fulfillment v0
  - `cancelShipmentOld`
  - `getAdditionalSellerInputsOld`
  - `getEligibleShipmentServicesOld`
- Notifications API: Notification Types
  - `LISTINGS_ITEM_ISSUES_CHANGE` (payload version 1.0)
  - `MFN_ORDER_STATUS_CHANGE`
  - `ORDER_STATUS_CHANGE`
- Product Pricing v2022-05-01
  - `CompetitivePriceThreshold` reference price type
- Reports API: [several](https://developer-docs.amazon.com/sp-api/docs/sp-api-deprecations#reports-api) reports
- Orders V0] Remove the Generic Orders Approvals workflow
  - `ItemApprovalTypes` and `ItemApprovalStatus` attributes

## Changed

### Version bumps
- [fulfillment-inbound-v0](https://developer-docs.amazon.com/sp-api/reference/fulfillment-inbound-v0) -> [fulfillment-inbound-v2024-03-20](https://developer-docs.amazon.com/sp-api/reference/fulfillment-inbound-v2024-03-20) (v0 supported under new namespace)
- [product-pricing-v0](https://developer-docs.amazon.com/sp-api/reference/product-pricing-v0) -> [productPricing_2022-05-01](https://developer-docs.amazon.com/sp-api/reference/product-pricing-v2022-05-01) (replaced old with new)
- [shippingV1](https://developer-docs.amazon.com/sp-api/docs/shipping-api-v1-reference) -> [shippingV2](https://developer-docs.shipping.amazon.com/apis/reference/getting-started) (replaced old with new)

## Added

### Domains
- [amazon-warehousing-and-distribution-model](https://developer-docs.amazon.com/sp-api/reference/amazon-warehousing-and-distribution-v2024-05-09)
- [application-integrations-api-model](https://developer-docs.amazon.com/sp-api/reference/app-integrations-v2024-04-01)
- [application-management-api-model](https://developer-docs.amazon.com/sp-api/reference/application-management-v2023-11-30)
- [customer-feedback-api-model](https://developer-docs.amazon.com/sp-api/reference/customer-feedback-v2024-06-01)
- [data-kiosk-api-model](https://developer-docs.amazon.com/sp-api/reference/data-kiosk-v2023-11-15)
- [delivery-by-amazon](https://developer-docs.amazon.com/sp-api/reference/delivery-by-amazon-v2022-07-01)
- [external-fulfillment](https://developer-docs.amazon.com/sp-api/reference/external-fulfillment-inventory-v2024-09-11)
- [fulfillment-inbound-api-model-V0](https://developer-docs.amazon.com/sp-api/reference/fulfillment-inbound-v0)
- [invoices-api-model](https://developer-docs.amazon.com/sp-api/reference/invoices-v2024-06-19)
- [replenishment-api-model](https://developer-docs.amazon.com/sp-api/reference/replenishment-v2022-11-07)
- [seller-wallet-api-model](https://developer-docs.amazon.com/sp-api/reference/seller-wallet-v2024-03-01)
- [supply-sources-api-model](https://developer-docs.amazon.com/sp-api/reference/supply-sources-v2020-07-01)
- [vehicles-api-model](https://developer-docs.amazon.com/sp-api/reference/vehicles-v2024-11-01)

### Domains with Subdomains
- finances-model _(likely causes an issue where only `transfers_2024` has a module replacing [finances-v0](https://developer-docs.amazon.com/sp-api/reference/finances-v0): this could be fixed in a follow up PR)_
  - [finances_2024-06-19](https://developer-docs.amazon.com/sp-api/reference/finances-v2024-06-19)
  - [transfers_2024-06-01](https://developer-docs.amazon.com/sp-api/reference/transfers-v2024-06-01)


# Methods

I found it helpful to compare the git trees to find changes at a dir level. Using git diff and then grouping took longer and was less accurate.
 
```
OLD=54f22853bf3937148461ec84a74eb8f58da13056
NEW=aad66aaf3225061891d194dfebcc9fa7697e8dd1

comm -12 \
  <(git ls-tree -d --name-only "$OLD:models" | sort) \
  <(git ls-tree -d --name-only "$NEW:models" | sort) \
| sed 's|^|models/|'
```
